### ISSUE TITLE
free(): invalid pointer when using pyvips with Torch

### ISSUE THREAD

**Issue by EKami** [conda-forge/libvips-feedstock#87]
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

The issue was initially reported [here](https://github.com/libvips/libvips/issues/3655) but according to the libvips authors and after my own investigation the issue only appears to affect the conda package.
Here is a detailed explanation of the issue:

### Description

If I try to load a Torchscript model but I import Pyvips before doing so, I get an invalid pointer exception.

### To reproduce

Use the following `Dockerfile`:
```
FROM ubuntu:22.04

RUN apt-get update \
 && apt-get install -y --no-install-recommends --allow-unauthenticated\
    nano \
    wget \
    curl \
    ca-certificates \
    apt-transport-https \
    build-essential \
    pkg-config \
    libglib2.0-dev \
    libexpat1-dev \
    libgirepository1.0-dev \
    python3 \
    python-is-python3 \
    python3-pip \
    python3-setuptools \
    python3-wheel \
    ninja-build \
    python3.10-dev \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*

RUN pip3 install meson
RUN wget -O /tmp/libvips.tar.xz https://github.com/libvips/libvips/releases/download/v8.14.4/vips-8.14.4.tar.xz
RUN cd /tmp && tar xf /tmp/libvips.tar.xz && cd /tmp/vips-8.14.4 \
   && meson setup build-dir \
   && cd build-dir \
   && ninja \
   && ninja install

RUN ldconfig
COPY ./save_model.py /root/
COPY ./load_model.py /root/

RUN pip3 install torch --index-url https://download.pytorch.org/whl/cpu
RUN pip3 install pyvips

# Do the same in a conda env
COPY ./conda.yml /root/
COPY ./install_micromamba.sh /root/

# Micromamba installation
RUN chmod +x /root/install_micromamba.sh && /root/install_micromamba.sh
ENV MAMBA_ROOT_PREFIX=/root/micromamba
RUN micromamba create --yes -f /root/conda.yml
```
with the following `install_micromamba.sh`:
```
#!/bin/bash

# This script is intended to be used inside the pw-ai docker image
set -o errexit
set -o pipefail
set -e

MICROMAMBA_URL=https://micro.mamba.pm/api/micromamba/linux-64/latest

echo "Installing mamba..."
curl -Ls $MICROMAMBA_URL | tar -xvj bin/micromamba

ln -s /usr/bin/micromamba /usr/bin/mamba
ln -s /usr/bin/micromamba /usr/bin/conda
mamba config append channels defaults

export MAMBA_ROOT_PREFIX=/root/micromamba

micromamba shell init --shell=bash --prefix=/root/micromamba
source /root/.bashrc
```
`conda.yml`:
```
name: test
channels:
  - conda-forge
  - pytorch
  - nvidia
dependencies:
 - pytorch=2.0.1
 - cpuonly
 - pyvips=2.2.1
 - libvips=8.14.4
```

`load_model.py`:
```
import pyvips
import torch

model = torch.jit.load("./mymodel.pt")
print("I should be displayed if I didn't crash! Otherwise comment out the pyvips import")
```

`save_model.py`:
```
import torch


class MyDecisionGate(torch.nn.Module):
    def forward(self, x):
        if x.sum() > 0:
            return x
        else:
            return -x

scripted_gate = torch.jit.script(MyDecisionGate())
scripted_gate.save("./mymodel.pt")
```

When I run the script in the system env it works fine but if I use the conda env I get:
```
root@1b7fb756efa6:~# micromamba activate test
(test) root@1b7fb756efa6:~# python load_model.py
free(): invalid pointer
Aborted (core dumped)
```

Using conda or micromamba gives the same result. 
Check the original post on the libvips repo, the authors seem to have clues on why this is happening.

### Installed packages

```shell
channels:
  - conda-forge
  - pytorch
  - nvidia
dependencies:
 - pytorch=2.0.1
 - cpuonly
 - pyvips=2.2.1
 - libvips=8.14.4
```

### Environment info

```shell
described in the Dockerfile
```

**Comment by hmaarrfk** [comment]
I can recreate with:
```
mamba create --name libvips87_mix --channel conda-forge --channel pytorch --channel nvidia --override-channels pytorch=2.0.1 cpuonly pyvips=2.2.1 libvips=8.14.4
conda activate libvips87_mix
```

I cannot recreate with
```
mamba create --name libvips87_cf_only --channel conda-forge  --override-channels pytorch=*=cpu* pyvips libvips
conda activate libvips87_cf_only
```

You are getting this error because you are mixing channels that are incompatible with conda-forge.

If you need this functionality, please raise an issue with pytorch and nvidia letting them know you want compatibility with conda-forge

**Comment by EKami** [comment]
I see, thank you, I did not know mixing up channels could result in segfaults/bad pointers like this.

**Comment by hmaarrfk** [comment]
yes, we should provide starker warnings.

we used to be compatible, which is why there is alot of confusion.

Why do you need their channels anyway? if you are on unix, we build pytorch fine.

**Comment by EKami** [comment]
This is the recommended way to install `pytorch` [on their website](https://pytorch.org/). I need Pytorch to work with CUDA which is not something that I can get if I install from `conda-forge`:
```
>>> import torch
>>> torch.cuda.is_available()
False
```
Moreover `conda-forge::pytorch` packages are often a bit behind in terms of versions.

**Comment by EKami** [comment]
So I'm in a situation where I need to choose one or the other but I can't get both in the same env because of the bad pointer =/.

**Comment by hmaarrfk** [comment]
> This is the recommended way to install pytorch [on their website](https://pytorch.org/). 

The would also recommend you stick to the python standard library.


> I need Pytorch to work with CUDA which is not something that I can get if I install from conda-forge

```
pytorch                   2.0.0           cuda112py39ha9981d0_200    conda-forge
```
exists from conda forge. it should just get detected automatically.

1. Remove the pytorch and nvidia channels
2. Instally `pytorch`
3. profit.

**Comment by EKami** [comment]
> 1. Remove the pytorch and nvidia channels
> 2. Instally `pytorch`
> 3. profit.

lol, like I said above this is exactly what I did, it does not support CUDA, it's no use to me.
With the following `conda.yaml`:
```
name: test
dependencies:
 - conda-forge::pytorch=*=cpu*
 - conda-forge::pyvips=2.2.1
 - conda-forge::libvips=8.14.4
 - conda-forge::xorg-libxau
```
And the same `Dockerfile` as above:
```
$ docker run -it --gpus all vips_test bash
(test) root@97dd41307e94:/# nvidia-smi
Mon Sep 11 02:29:41 2023
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0  On |                  N/A |
| N/A   45C    P8     8W / 175W |      1MiB / 16376MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
root@97dd41307e94:/# micromamba activate test
(test) root@97dd41307e94:/# python
Python 3.11.5 | packaged by conda-forge | (main, Aug 27 2023, 03:34:09) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
False
```

**Comment by hmaarrfk** [comment]
is there a reason you are adding `*=cpu*`? to the end of the pytorch package?

**Comment by EKami** [comment]
oh wait, my bad. Let me try again

**Comment by hmaarrfk** [comment]
```
$ mamba list pytorch
# packages in environment at /home/mark/mambaforge/envs/dev:
#
# Name                    Version                   Build  Channel
libopenvino-pytorch-frontend 2023.0.1             h59595ed_2    conda-forge
pytorch                   2.0.0           cuda112py39ha9981d0_200    conda-forge
$ nvidia-smi 
Sun Sep 10 22:32:07 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.05              Driver Version: 535.86.05    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4090        Off | 00000000:42:00.0 Off |                  Off |
|  0%   37C    P8              10W / 450W |    130MiB / 24564MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      1593      G   /usr/lib/xorg/Xorg                           56MiB |
|    0   N/A  N/A      1988      G   /usr/bin/gnome-shell                         60MiB |
+---------------------------------------------------------------------------------------+
$ python -c "import torch; print( torch.cuda.is_available())"
True
```

**Comment by EKami** [comment]
Thanks! Now I need to test if the cuda 11.2 version from `conda-forge` doesn't conflict with CUDA used by Tensorflow in the same env lol

**Comment by hmaarrfk** [comment]
again, just download tensorflow from us.

you'll want to evetually get rid of both packages, they are quite hard to maintain.

good luck

**Comment by EKami** [comment]
Do you know how I can force the cuda package to be pulled inside the `cuda.yaml` file?
Using this:
```
name: test
dependencies:
 - conda-forge::pytorch
 - conda-forge::pyvips=2.2.1
 - conda-forge::libvips=8.14.4
 - conda-forge::xorg-libxau
 ```
 Gives me:
 ```
 (test) root@6dc4e6a82920:/# mamba list pytorch
List of packages in environment: "/root/micromamba/envs/test"

  Name     Version  Build                      Channel
────────────────────────────────────────────────────────────
  pytorch  2.0.0    cpu_mkl_py311had667d7_101  conda-forge
 ```

**Comment by EKami** [comment]
with `conda-forge::pytorch=*=cuda*` I guess. Thanks!

**Comment by hmaarrfk** [comment]
are you building in docker?

Using `conda-forge::` means you didn't setup conda-forge as the main (and only!) channel.

Once you have your conda setup, you shouldn't need to use `conda-forge::` everywhere....


https://conda-forge.org/docs/user/tipsandtricks.html#installing-cuda-enabled-packages-like-tensorflow-and-pytorch

**Comment by EKami** [comment]
Both in and outside docker yes (the most important is within docker). I fixed the `conda.yaml` and now it's working:
```
name: test
channels:
 - conda-forge
dependencies:
 - pytorch=*=cuda*
 - pyvips=2.2.1
 - libvips=8.14.4
 - xorg-libxau
```

I guess it didn't work previously because I was strict on channel selection and no other package outside my `conda.yaml` file could be pulled.
Thanks a lot for your help!!

**Comment by hmaarrfk** [comment]
we don't have 2.0.1, that might have been the issue

**Comment by EKami** [comment]
Looks like I still won't be able to use `conda-forge` only tho :( :
```
error    libmamba Could not solve for environment specs
    The following packages are incompatible
    ├─ pyg   is installable with the potential options
    │  ├─ pyg [2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ pytorch-cuda 11.7.* , which does not exist (perhaps a missing channel);
    │  ├─ pyg [2.0.4|2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ python >=3.8,<3.9.0a0 , which can be installed;
    │  ├─ pyg [2.0.0|2.0.1|...|2.3.1] would require
    │  │  └─ cpuonly  , which does not exist (perhaps a missing channel);
    │  ├─ pyg [2.0.0|2.0.1|...|2.3.1] would require
    │  │  └─ python_abi 3.8.* *_cp38, which can be installed;
    │  ├─ pyg [2.0.4|2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ python >=3.7,<3.8.0a0 , which can be installed;
    │  ├─ pyg [2.0.4|2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ python >=3.9,<3.10.0a0 , which can be installed;
    │  ├─ pyg [2.0.0|2.0.1|...|2.3.1] would require
    │  │  └─ python_abi 3.7.* *_cp37m, which can be installed;
    │  ├─ pyg [2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ python >=3.10,<3.11.0a0 , which can be installed;
    │  ├─ pyg [2.0.0|2.0.1|2.0.2|2.0.3] would require
    │  │  └─ python_abi 3.6.* *_cp36m, which can be installed;
    │  ├─ pyg [2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ pytorch-cuda 11.6.* , which does not exist (perhaps a missing channel);
    │  ├─ pyg [2.0.0|2.0.1|...|2.3.1] would require
    │  │  └─ python_abi 3.9.* *_cp39, which can be installed;
    │  ├─ pyg [2.1.0|2.2.0|2.3.0|2.3.1] would require
    │  │  └─ python_abi 3.10.* *_cp310, which can be installed;
    │  └─ pyg [2.3.0|2.3.1] would require
    │     └─ pytorch-cuda 11.8.* , which does not exist (perhaps a missing channel);
    └─ python 3.11**  is uninstallable because it requires
       └─ python_abi 3.11.* *_cp311, which conflicts with any installable versions previously reported.
critical libmamba Could not solve for environment specs
```

I have this:
```
name: dl
channels:
 # Only use conda-forge to avoid conflicts, packages not from this channel should be specified in
 # dependencies
 - conda-forge
dependencies:
 - python=3.11
 # Important to avoid the white patches bug
 - pixman=0.40.0
 - boto3
 - checksumdir
 - click
 - conda
 - dacite
 - docker-py
 - gitpython
 - h5py
 - imagehash
 - imagemagick
 - jsonschema
 - faiss-gpu
 - keras
 - keras-preprocessing
 - lxml
 - matplotlib
 - numpy
 - opencv
 - openslide
 - openslide-python
 - pandas
 - pillow
 - pip
 - pyrsistent
 - protobuf
 - py-opencv
 - pyarrow
 - pypdf3
 - pytest
 - pytest-metadata
 - pytest-html
 - python-dateutil
 - python-dotenv
 - pyvips
 - pyyaml
 - polars
 - requests
 - scikit-image
 - scikit-learn
 - scipy
 - seaborn
 - sentry-sdk
 - tabulate
 - tqdm
 - tensorflow-gpu=2.12.1
 - wandb
 - xhtml2pdf
 - zstandard
 - pytest-forked
 - pytorch=2.0.0
 - xorg-libxau
# Non conda-forge packages
 - pyg::pyg
```

```
$ micromamba search -c conda-forge pyg
Getting repodata from channels...

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache


No entries matching "pyg" found
Channels may not be configured. Try giving a channel with '-c,--channel' option, or use `--use-local=1` to search for installed packages.
```

**Comment by hmaarrfk** [comment]
python 3.11 is "new"
https://conda-forge.org/status/#python311

**Comment by EKami** [comment]
I don't think that's the issue. This doesn't work either:
```
name: dl2
channels:
 - conda-forge
dependencies:
 - python=3.10
 # Important to avoid the white patches bug
 - pixman=0.40.0
 - boto3
 - checksumdir
 - click
 - conda
 - dacite
 - docker-py
 - gitpython
 - h5py
 - imagehash
 - imagemagick
 - jsonschema
 - faiss-gpu
 - keras
 - keras-preprocessing
 - lxml
 - matplotlib
 - numpy
 - opencv
 - openslide
 - openslide-python
 - pandas
 - pillow
 - pip
 - pyrsistent
 - protobuf
 - py-opencv
 - pyarrow
 - pypdf3
 - pytest
 - pytest-metadata
 - pytest-html
 - python-dateutil
 - python-dotenv
 - pyvips
 - pyyaml
 - polars
 - requests
 - scikit-image
 - scikit-learn
 - scipy
 - seaborn
 - sentry-sdk
 - tabulate
 - tqdm
 - tensorflow-gpu=2.12.1
 - wandb
 - xhtml2pdf
 - zstandard
 - pytest-forked
 - pytorch=2.0.0
 - xorg-libxau
# Non conda-forge packages
# Even pyg:pyg doesn't work because it relies on pytorch channel
 - pyg
 ```
 It's just that `pyg` doesn't seem to exist on `conda-forge` and the `pyg::pyg` package relies on the one from the `pytorch` channel.
So I'm back at square one doing this:
```
name: dl
channels:
 - conda-forge
dependencies:
 - python=3.11
 # Important to avoid the white patches bug
 - pixman=0.40.0
 - boto3
 - checksumdir
 - click
 - conda
 - dacite
 - docker-py
 - gitpython
 - h5py
 - imagehash
 - imagemagick
 - jsonschema
 - faiss-gpu
 - keras
 - keras-preprocessing
 - lxml
 - matplotlib
 - numpy
 - opencv
 - openslide
 - openslide-python
 - pandas
 - pillow
 - pip
 - pyrsistent
 - protobuf
 - py-opencv
 - pyarrow
 - pypdf3
 - pytest
 - pytest-metadata
 - pytest-html
 - python-dateutil
 - python-dotenv
 - pyvips
 - pyyaml
 - polars
 - requests
 - scikit-image
 - scikit-learn
 - scipy
 - seaborn
 - sentry-sdk
 - tabulate
 - tqdm
 - tensorflow-gpu=2.12.1
 - wandb
 - xhtml2pdf
 - zstandard
 - pytest-forked
# Non conda-forge packages
 - pytorch::pytorch=2.0.1
 - pytorch::pytorch-cuda=11.8
 - nvidia::cuda-cudart
 - pyg::pyg
```
Which gives me the segfault :(

**Comment by hmaarrfk** [comment]
hmmm, basically,

1. pytorch channel first called their package `pytorch-gpu`
2. then they changed their mind and just removed the suffixes
3. then now they changed their mind and called it `pytorch-cuda`

so the `pyg::pyg` channel requires a package called `pytorch-cuda` which we don't build. `pytorch-cuda` is the same as our `pytorch-gpu` package. just helps get the GPU build

**Comment by hmaarrfk** [comment]
https://github.com/conda-forge/pytorch-cpu-feedstock/issues/187

You have a few choices:

1. Use a combination of `--no-deps` and install things manually. This is somewhat ok for docker since you are always rebuilding.
2. Ask the team at pyg to remove the duplicated effort for cpuonly and -cuda, the package is the same
3. Build `pyg` for conda-forge
4. Use pip
5. something else


goodluck

**Comment by EKami** [comment]
Thanks a lot for your help!
Yeah, I'll figure something out... eventually lol.
Is there any reason why `pyg` wasn't selected as a package to build on `conda-forge`?
How do you select what packages to build?

**Comment by hmaarrfk** [comment]
https://conda-forge.org/docs/maintainer/adding_pkgs.html
